### PR TITLE
Add Consumer Group abstraction

### PR DESCRIPTION
**Pull request by octosteve**
# Consumer Groups
Failure is hard. This PR introduces a way to group consumers you _want_ to fail together.

## Usage
To Group a set of consumers together You would add this to your supervision tree
```elixir
children = [
#... more things
  RailwayIpc.ConsumerGroup.Supervisor.setup_group(children:[Consumer1, Consumer2],
    name: :unique_name_for_this_consumer_group)
# start your supervisor
```
This will make it so that ANY of these consumers crashing, brings down ONLY other consumers in this consumer group.

## When would you need this?
There are consumers in our system where failure could lead to either inconsistent data being sent to 3rd party systems or failures in different consumers "at a distance". A failure to integrate a new Campus into our DB may not break ALL new Cohorts, but it will as soon as someone tries to add Cohort for that Campus. 

Other times, failures may not affect other parts of the application if they fail. For instance, we can allow for a failure to sync Cohort information to SFDC to only fail Consumers that send information to SFDC, but leave the ones that register students to cohort alone.

Previous approaches to failure were heavy handed and stopped ALL consumers. This works if the application is small and has one purpose. With larger applications however we may have several disparate concerns that can continue to operate even if one part of the system is dealing with an outage.

# Restarting the tree
In the event that the reason your consumer group crashed was a _bad message_, simply remove the bad message from the queue, and restart your tree using `RailwayIpc.ConsumerGroup.Supervisor.restart_tree(:unique_name_for_this_consumer_group)`. This allows you to restart workers without having to deploy. The application a new.

Here's an example of isolated groups only crashing their peers.
failure_groups
Here's an example of how you'd set this up:
image

### CODE REVIEW

File: lib/railway_ipc/consumer_group/consumer_supervisor.ex [hunk 1]
```diff
@@ -0,0 +1,21 @@
+defmodule RailwayIpc.ConsumerGroup.ConsumerSupervisor do
```
Comment (aturkewi) [inline lib/railway_ipc/consumer_group/consumer_supervisor.ex:1]: What's the difference between the ConsumerSupervisor and  the Supervisor?

**Review by aturkewi — COMMENTED** [review]
I'm a little lost here. There seems to be a lot going on and I don't think I really understand the structure that this builds with the two layers of supervisors (am I reading this correctly at all?).
It sounds like what you're building is the `:one_for_all` strategy though. So if one consumer in the group goes down, it takes the rest down. Why not have each app use that? I don't mean this as a suggestion, but maybe understanding why that path doesn't work can help me better understand the added complexity of this solution.

File: lib/railway_ipc/consumer_group/consumer_supervisor.ex (hunk 1 above)
Comment (octosteve) [reply lib/railway_ipc/consumer_group/consumer_supervisor.ex:1]: image
We're building something like that. The Supervisor is the parent for both the consumer supervisor and the monitor. This lets you restart the tree from the parent if needed.

Comment (octosteve) [reply lib/railway_ipc/consumer_group/consumer_supervisor.ex:1]: The consumer supervisor is just a supervisor that starts _your_ consumers.

**Comment by octosteve** [comment]
> I'm a little lost here. There seems to be a lot going on and I don't think I really understand the structure that this builds with the two layers of supervisors (am I reading this correctly at all?).
> It sounds like what you're building is the `:one_for_all` strategy though. So if one consumer in the group goes down, it takes the rest down. Why not have each app use that? I don't mean this as a suggestion, but maybe understanding why that path doesn't work can help me better understand the added complexity of this solution.
One thing to note is that we don't want to restart the tree automatically. If one dies, we want them _all_ to die and stay dead until we either restart it manually, or deploy to fix it. One for All _will_ kill all of the children and restart them if the strategy is the default `:permanent` but it doesn't do anything if the strategy is `:temporary`, which the ex_rabbit_pool pr makes them.
This is worth a read: https://hostiledeveloper.com/2020/08/12/superb-supervisors-designing-for-failure.html (Sorry for the homework)

**Comment by aturkewi** [comment]
Thanks for the explanation! And no worries about the homework :) I asked for it because I want to be able to understand what's going on here.

**Review by aturkewi — APPROVED** [review]
This looks good to me!
Thanks for adding!!

**Review by bnadlerjr — APPROVED** [review]

**Pull request merged**